### ISSUE TITLE
Include Selectors in Terminal

### ISSUE THREAD

**Issue by mottosso** [pyblish/pyblish-qml#64]
# Goal

To include results from processing Selectors in the Terminal so as to facilitate debugging.

Previously, Selectors were run on the host prior to launching the GUI. The GUI would then reap the selected instances and send them across for processing by subsequent plug-ins, like Validators.

However, as Selectors are equally important to get right and are as capable of producing error or messages of it's own, it makes sense for them to also take part of the growing error visualisation generated by Pyblish QML.

# Quirks

There are a few quirks about the approach currently.

### Server Compatibility 

Plug-in/Instance compatibility are currently computed within the host, right after Selection has taken place. The host then has access to all plug-ins and instances and can run them up against each other to find which are compatible. With this approach, Selection would need to happen from the GUI which means the host would no longer have access to all instances; only plug-ins.

Solution; performing filtering within the GUI.

### Double Trip

Currently, a single call to a host is made, asking it to deliver Plug-ins and Selected instances. We'll now have to make two calls; one for Plug-ins and one for Instances.

Practically speaking however this is probably in our favour as it means the GUI may show up before Selection has completed. In cases where Selection is an intensive process, it'll make for a better user experience.

**Comment by mottosso** [comment]
Implemented

![untitled](https://cloud.githubusercontent.com/assets/2152766/6831544/db2bb3be-d318-11e4-9f46-50bf66973daa.gif)

**Comment by mottosso** [comment]
## Client-Side Compatibility

In the above illustration, we can see that we're getting *all* plug-ins, even `ValidateIsIncompatible` which doesn't have any compatible instances. Currently, this should not have been there as it would have been filtered out so as to only show plug-ins that would actually take part of a publish.

However, because we won't be getting instances until after having processed each Selector, we can't know which to exclude until it's done. Which means we're left with two choices.

1. Show all plug-ins, and when selection is done remove plug-ins that didn't get any compatible instance(s)
2. Hide everything until Selection is complete.

Hiding everything is possibly more intuitive, as it might get confusing seeing some plug-ins vanish. However, it would also mean that in cases where Selection takes a substantial amount of time, there would be nothing to look at until it has completed.

Practically, this would only be relevant in cases where Selection takes longer to finish than the GUI takes to get drawn (> 100 ms).

**Comment by davidmartinezanim** [comment]
:+1: 

I'd personally go for option 2.

**Comment by tokejepsen** [comment]
Doesnt it mean that we need a two step process, so either pressing the publish button twice or have a separate button for selectors?

**Comment by mottosso** [comment]
> Doesnt it mean that we need a two step process, so either pressing the publish button twice or have a separate button for selectors?

The Reset button is running Selection. It's doing that now too, but it doesn't get visualised like it does in the Gif above, this shouldn't change that behaviour.

**Comment by tokejepsen** [comment]
> The Reset button is running Selection. It's doing that now too, but it doesn't get visualised like it does in the Gif above, this shouldn't change that behaviour.

Thats alright then:) option 2 for me as well

**Comment by mottosso** [comment]
Agreed. :)

Could be a nice spot for a loading notification/animation of sorts, possibly a progress bar.

**Comment by mottosso** [comment]
Client-side exclusion of non-compatible plug-ins works well with this.
https://gist.github.com/mottosso/a55f42fe3f184369eefd

```python
# Snippet
plugin_proxy.addExclusion("hasCompatible", False)
# hasCompatible is computed after selection
```

**Comment by mottosso** [comment]
Implemented in 0.2.7
